File: Backend/crates/modules/flora-economy-crypto/src/ledger.rs
//! Журнал FEP: append-only хеш-цепочка экономических событий (класс Certificate Transparency,
//! FGP §6.4 / FGP-CRYPTO §8).
//!
//! Каждая запись ссылается на хеш предыдущей; head подписывают независимые витнессы
//! (косайнинг — уровень модуля/инфраструктуры, ядро даёт канонические байты STH). Любой клиент
//! (wasm-сборка этого же crate в `@flora/client-core`) может воспроизвести состояние реплеем
//! журнала и сверить хеши — «не верь серверу, проверь сам».
//!
//! В журнале **нет** плутократического консенсуса: порядок задаёт оператор-секвенсор, а честность
//! секвенсора проверяема (реплей + витнессы + внешнее якорение). Компрометация секвенсора ⇒
//! расхождение реплея у первого же клиента ⇒ событие лестницы деградации FGP §7.3 — тот же
//! паттерн отказобезопасности, что у governance-журнала.

use serde::{Deserialize, Serialize};

use crate::amount::{AccountId, Grains, Timestamp};
use crate::canonical::CanonicalWriter;
use crate::domain;
use crate::hash::{Hash32, ZERO_HASH, tagged};
use crate::params::Parameters;
use crate::sig::{PublicKeyBytes, SignatureBytes};

/// Тело записи журнала — все экономические события FEP v1.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "kind",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum EntryBody {
    /// Учреждение журнала: версия протокола + genesis-параметры. Создаёт **ноль** grain.
    Genesis {
        protocol_version: u16,
        params: Parameters,
    },
    /// Смена параметров решением governance (R2); `policy_ref` — ссылка на PolicyArtifact FGP.
    ParametersUpdated {
        params: Parameters,
        policy_ref: String,
    },
    /// Открытие экономического аккаунта (владелец — человек или commons-структура).
    AccountOpened {
        account: AccountId,
        /// Ed25519-ключ владения аккаунтом (авторизация переводов).
        #[serde(with = "crate::hexser")]
        owner_key: PublicKeyBytes,
    },
    /// Начисление UBI за диапазон эпох. Единственный источник эмиссии.
    UbiIssued {
        account: AccountId,
        from_epoch: u64,
        to_epoch: u64,
        amount: Grains,
    },
    /// Начисление демерреджа: списание с аккаунта в Commons-казну.
    DemurrageCharged {
        account: AccountId,
        periods: u64,
        amount: Grains,
    },
    /// Перевод Pollen. Подпись отправителя обязательна.
    Transfer {
        from: AccountId,
        to: AccountId,
        amount: Grains,
        /// Клиентский идемпотентный идентификатор (16 байт).
        #[serde(with = "crate::hexser")]
        nonce: [u8; 16],
        #[serde(with = "crate::hexser")]
        signature: SignatureBytes,
    },
    /// Открытие/изменение линии доверия (лимиты со стороны каждого участника).
    TrustlineSet {
        lo: AccountId,
        hi: AccountId,
        limit_lo_to_hi: Grains,
        limit_hi_to_lo: Grains,
        #[serde(with = "crate::hexser")]
        signature_lo: SignatureBytes,
        #[serde(with = "crate::hexser")]
        signature_hi: SignatureBytes,
    },
    /// Платёж по цепочке линий доверия (path[0] — плательщик, последний — получатель).
    CreditTransfer {
        path: Vec<AccountId>,
        amount: Grains,
        #[serde(with = "crate::hexser")]
        nonce: [u8; 16],
        #[serde(with = "crate::hexser")]
        signature: SignatureBytes,
    },
    /// Расход Commons-казны по ратифицированной категории бюджета (FGP §10.2):
    /// казна платит получателю; `policy_ref` — категория действующего бюджет-артефакта.
    CommonsSpend {
        to: AccountId,
        amount: Grains,
        policy_ref: String,
    },
}

impl EntryBody {
    /// Канонические байты тела (consensus-путь: вход хеша записи).
    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut w = CanonicalWriter::new();
        match self {
            EntryBody::Genesis {
                protocol_version,
                params,
            } => {
                w.u8(0).u16(*protocol_version);
                write_params(&mut w, params);
            }
            EntryBody::ParametersUpdated { params, policy_ref } => {
                w.u8(1);
                write_params(&mut w, params);
                w.str(policy_ref);
            }
            EntryBody::AccountOpened { account, owner_key } => {
                w.u8(2).account(account).bytes(owner_key);
            }
            EntryBody::UbiIssued {
                account,
                from_epoch,
                to_epoch,
                amount,
            } => {
                w.u8(3)
                    .account(account)
                    .u64(*from_epoch)
                    .u64(*to_epoch)
                    .grains(*amount);
            }
            EntryBody::DemurrageCharged {
                account,
                periods,
                amount,
            } => {
                w.u8(4).account(account).u64(*periods).grains(*amount);
            }
            EntryBody::Transfer {
                from,
                to,
                amount,
                nonce,
                signature,
            } => {
                w.u8(5)
                    .account(from)
                    .account(to)
                    .grains(*amount)
                    .bytes(nonce)
                    .bytes(signature);
            }
            EntryBody::TrustlineSet {
                lo,
                hi,
                limit_lo_to_hi,
                limit_hi_to_lo,
                signature_lo,
                signature_hi,
            } => {
                w.u8(6)
                    .account(lo)
                    .account(hi)
                    .grains(*limit_lo_to_hi)
                    .grains(*limit_hi_to_lo)
                    .bytes(signature_lo)
                    .bytes(signature_hi);
            }
            EntryBody::CreditTransfer {
                path,
                amount,
                nonce,
                signature,
            } => {
                w.u8(7)
                    .account_list(path)
                    .grains(*amount)
                    .bytes(nonce)
                    .bytes(signature);
            }
            EntryBody::CommonsSpend {
                to,
                amount,
                policy_ref,
            } => {
                w.u8(8).account(to).grains(*amount).str(policy_ref);
            }
        }
        w.finish()
    }
}

fn write_params(w: &mut CanonicalWriter, p: &Parameters) {
    w.u32(p.demurrage_ppm_per_period)
        .i64(p.demurrage_period_ms)
        .grains(p.demurrage_exempt_threshold)
        .grains(p.ubi_per_epoch)
        .i64(p.ubi_epoch_ms)
        .u64(p.ubi_max_backfill_epochs)
        .grains(p.trustline_max_limit)
        .u8(p.credit_path_max_hops);
}

/// Запись журнала: порядковый номер, время, тело и хеш-ссылка на предыдущую.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LedgerEntry {
    /// Порядковый номер (genesis = 0).
    pub seq: u64,
    /// Метка времени секвенсора (Unix-мс UTC); монотонна по журналу.
    pub at: Timestamp,
    /// Хеш предыдущей записи ([`ZERO_HASH`] для genesis).
    #[serde(with = "crate::hexser")]
    pub prev_hash: Hash32,
    /// Тело события.
    pub body: EntryBody,
}

impl LedgerEntry {
    /// Хеш записи: `H(leaf-label ‖ seq ‖ at ‖ prev_hash ‖ body_bytes)`.
    pub fn entry_hash(&self) -> Hash32 {
        let mut w = CanonicalWriter::new();
        w.u64(self.seq)
            .timestamp(self.at)
            .hash(&self.prev_hash)
            .bytes(&self.body.canonical_bytes());
        tagged(domain::LEDGER_LEAF, w.as_slice())
    }
}

/// Head журнала (аналог Signed Tree Head CT): всё, что подписывают витнессы.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LedgerHead {
    /// Число записей в журнале.
    pub size: u64,
    /// Хеш последней записи (хеш-цепочка).
    #[serde(with = "crate::hexser")]
    pub last_entry_hash: Hash32,
    /// Merkle-корень всех entry_hash (для inclusion/consistency-доказательств).
    #[serde(with = "crate::hexser")]
    pub merkle_root: Hash32,
    /// Метка времени последней записи.
    pub at: Timestamp,
}

impl LedgerHead {
    /// Канонические байты head — вход подписи витнесса (`LEDGER_STH`).
    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut w = CanonicalWriter::new();
        w.u64(self.size)
            .hash(&self.last_entry_hash)
            .hash(&self.merkle_root)
            .timestamp(self.at);
        w.finish()
    }

    /// Head пустого журнала (до genesis).
    pub fn empty() -> LedgerHead {
        LedgerHead {
            size: 0,
            last_entry_hash: ZERO_HASH,
            merkle_root: ZERO_HASH,
            at: Timestamp(0),
        }
    }
}

/// Канонические байты авторизации перевода (то, что подписывает отправитель).
pub fn transfer_signing_bytes(
    from: &AccountId,
    to: &AccountId,
    amount: Grains,
    nonce: &[u8; 16],
) -> Vec<u8> {
    let mut w = CanonicalWriter::new();
    w.account(from).account(to).grains(amount).bytes(nonce);
    w.finish()
}

/// Канонические байты авторизации линии доверия (подписывают обе стороны).
pub fn trustline_signing_bytes(
    lo: &AccountId,
    hi: &AccountId,
    limit_lo_to_hi: Grains,
    limit_hi_to_lo: Grains,
) -> Vec<u8> {
    let mut w = CanonicalWriter::new();
    w.account(lo)
        .account(hi)
        .grains(limit_lo_to_hi)
        .grains(limit_hi_to_lo);
    w.finish()
}

/// Канонические байты авторизации перевода по взаимному кредиту (подписывает плательщик).
pub fn credit_transfer_signing_bytes(
    path: &[AccountId],
    amount: Grains,
    nonce: &[u8; 16],
) -> Vec<u8> {
    let mut w = CanonicalWriter::new();
    w.account_list(path).grains(amount).bytes(nonce);
    w.finish()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn acc(n: u8) -> AccountId {
        AccountId([n; 16])
    }

    #[test]
    fn entry_hash_is_deterministic_and_sensitive() {
        let entry = LedgerEntry {
            seq: 0,
            at: Timestamp(1_000),
            prev_hash: ZERO_HASH,
            body: EntryBody::Genesis {
                protocol_version: 1,
                params: Parameters::genesis(),
            },
        };
        assert_eq!(entry.entry_hash(), entry.entry_hash());

        let mut other = entry.clone();
        other.at = Timestamp(1_001);
        assert_ne!(entry.entry_hash(), other.entry_hash());
    }

    #[test]
    fn body_variants_have_distinct_canonical_bytes() {
        let a = EntryBody::UbiIssued {
            account: acc(1),
            from_epoch: 0,
            to_epoch: 0,
            amount: Grains(5),
        }
        .canonical_bytes();
        let b = EntryBody::DemurrageCharged {
            account: acc(1),
            periods: 0,
            amount: Grains(5),
        }
        .canonical_bytes();
        assert_ne!(a, b, "разные варианты обязаны иметь разные байты (u8-тег)");
    }

    #[test]
    fn signing_bytes_bind_all_fields() {
        let base = transfer_signing_bytes(&acc(1), &acc(2), Grains(10), &[0u8; 16]);
        assert_ne!(
            base,
            transfer_signing_bytes(&acc(1), &acc(2), Grains(11), &[0u8; 16])
        );
        assert_ne!(
            base,
            transfer_signing_bytes(&acc(1), &acc(3), Grains(10), &[0u8; 16])
        );
        assert_ne!(
            base,
            transfer_signing_bytes(&acc(1), &acc(2), Grains(10), &[1u8; 16])
        );
    }

    #[test]
    fn json_serialization_is_camel_case() {
        let entry = LedgerEntry {
            seq: 3,
            at: Timestamp(5),
            prev_hash: ZERO_HASH,
            body: EntryBody::CommonsSpend {
                to: acc(2),
                amount: Grains(7),
                policy_ref: "budget-2026/infrastructure".into(),
            },
        };
        let json = serde_json::to_string(&entry).unwrap();
        assert!(json.contains("\"prevHash\""));
        assert!(json.contains("\"policyRef\""));
        assert!(json.contains("\"kind\":\"commonsSpend\""));
    }
}

File: Packages/flora-client-core/src/economy/ledger.ts
/**
 * Журнал FEP на клиенте: типы JSON-представления (контракт JSONL-журнала и HTTP API),
 * канонические байты и хеши записей — зеркало `flora-economy-crypto::ledger`.
 *
 * JSON (camelCase, hex для байтовых полей, суммы — grain числом) — только транспорт;
 * в хеши входит **канонический бинарный формат** (FEP.md §9.3). Параметры протокола
 * сериализуются snake_case — как в Rust-ядре (контракт журнала).
 */

import { CanonicalWriter } from "./canonical.js";
import { FEP_LEDGER_LEAF } from "./domainTags.js";
import { EconomyCodecError, fromHex, toHex } from "./encoding.js";
import { sha256Tagged, type Hash32 } from "./hash.js";

// ---------- типы JSON-контракта ----------

/** Экономические параметры (FEP.md, Приложение A). В JSON журнала — snake_case. */
export type EconomyParametersJson = {
  demurrage_ppm_per_period: number;
  demurrage_period_ms: number;
  demurrage_exempt_threshold: number;
  ubi_per_epoch: number;
  ubi_epoch_ms: number;
  ubi_max_backfill_epochs: number;
  trustline_max_limit: number;
  credit_path_max_hops: number;
};

/** AccountId в JSON журнала — массив из 16 байт (сырые байты UUID). */
export type AccountIdJson = number[];

/** Тело записи журнала — все экономические события FEP v1 (`kind`-тег, camelCase). */
export type LedgerEntryBodyJson =
  | { kind: "genesis"; protocolVersion: number; params: EconomyParametersJson }
  | { kind: "parametersUpdated"; params: EconomyParametersJson; policyRef: string }
  | { kind: "accountOpened"; account: AccountIdJson; ownerKey: string }
  | {
      kind: "ubiIssued";
      account: AccountIdJson;
      fromEpoch: number;
      toEpoch: number;
      amount: number;
    }
  | { kind: "demurrageCharged"; account: AccountIdJson; periods: number; amount: number }
  | {
      kind: "transfer";
      from: AccountIdJson;
      to: AccountIdJson;
      amount: number;
      nonce: string;
      signature: string;
    }
  | {
      kind: "trustlineSet";
      lo: AccountIdJson;
      hi: AccountIdJson;
      limitLoToHi: number;
      limitHiToLo: number;
      signatureLo: string;
      signatureHi: string;
    }
  | {
      kind: "creditTransfer";
      path: AccountIdJson[];
      amount: number;
      nonce: string;
      signature: string;
    }
  | { kind: "commonsSpend"; to: AccountIdJson; amount: number; policyRef: string };

/** Запись журнала: порядковый номер, время, хеш-ссылка на предыдущую, тело. */
export type LedgerEntryJson = {
  seq: number;
  at: number;
  prevHash: string;
  body: LedgerEntryBodyJson;
};

/** Head журнала (Signed Tree Head) — то, что подписывают витнессы. */
export type LedgerHeadJson = {
  size: number;
  lastEntryHash: string;
  merkleRoot: string;
  at: number;
};

// ---------- аккаунты ----------

/** JSON-массив байт → 16-байтовый AccountId. */
export function accountBytesFromJson(value: AccountIdJson): Uint8Array {
  if (value.length !== 16) {
    throw new EconomyCodecError(`AccountId обязан быть 16 байт, получено ${value.length}`);
  }
  const out = new Uint8Array(16);
  for (let i = 0; i < 16; i += 1) {
    const byte = value[i];
    if (typeof byte !== "number" || !Number.isInteger(byte) || byte < 0 || byte > 255) {
      throw new EconomyCodecError(`AccountId: байт ${i} вне диапазона`);
    }
    out[i] = byte;
  }
  return out;
}

/** 16 байт → JSON-представление AccountId (массив чисел). */
export function accountJsonFromBytes(bytes: Uint8Array): AccountIdJson {
  if (bytes.length !== 16) {
    throw new EconomyCodecError(`AccountId обязан быть 16 байт, получено ${bytes.length}`);
  }
  return Array.from(bytes);
}

/** UUID (с дефисами или 32 hex-символа) → 16 байт AccountId. Аккаунт = UUID пользователя. */
export function accountBytesFromUuid(uuid: string): Uint8Array {
  const compact = uuid.replaceAll("-", "").toLowerCase();
  if (compact.length !== 32) {
    throw new EconomyCodecError(`некорректный UUID аккаунта: ${uuid}`);
  }
  return fromHex(compact, 16);
}

/** 16 байт AccountId → канонический UUID lowercase с дефисами. */
export function uuidFromAccountBytes(bytes: Uint8Array): string {
  if (bytes.length !== 16) {
    throw new EconomyCodecError(`AccountId обязан быть 16 байт, получено ${bytes.length}`);
  }
  const hex = toHex(bytes);
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

// ---------- канонические байты ----------

function writeParams(w: CanonicalWriter, p: EconomyParametersJson): void {
  w.u32(p.demurrage_ppm_per_period)
    .i64(p.demurrage_period_ms)
    .grains(BigInt(p.demurrage_exempt_threshold))
    .grains(BigInt(p.ubi_per_epoch))
    .i64(p.ubi_epoch_ms)
    .u64(p.ubi_max_backfill_epochs)
    .grains(BigInt(p.trustline_max_limit))
    .u8(p.credit_path_max_hops);
}

/** Канонические байты тела записи (consensus-путь: вход хеша записи). */
export function entryBodyCanonicalBytes(body: LedgerEntryBodyJson): Uint8Array {
  const w = new CanonicalWriter();
  switch (body.kind) {
    case "genesis": {
      w.u8(0).u16(body.protocolVersion);
      writeParams(w, body.params);
      break;
    }
    case "parametersUpdated": {
      w.u8(1);
      writeParams(w, body.params);
      w.str(body.policyRef);
      break;
    }
    case "accountOpened": {
      w.u8(2).account(accountBytesFromJson(body.account)).bytes(fromHex(body.ownerKey, 32));
      break;
    }
    case "ubiIssued": {
      w.u8(3)
        .account(accountBytesFromJson(body.account))
        .u64(body.fromEpoch)
        .u64(body.toEpoch)
        .grains(BigInt(body.amount));
      break;
    }
    case "demurrageCharged": {
      w.u8(4)
        .account(accountBytesFromJson(body.account))
        .u64(body.periods)
        .grains(BigInt(body.amount));
      break;
    }
    case "transfer": {
      w.u8(5)
        .account(accountBytesFromJson(body.from))
        .account(accountBytesFromJson(body.to))
        .grains(BigInt(body.amount))
        .bytes(fromHex(body.nonce, 16))
        .bytes(fromHex(body.signature, 64));
      break;
    }
    case "trustlineSet": {
      w.u8(6)
        .account(accountBytesFromJson(body.lo))
        .account(accountBytesFromJson(body.hi))
        .grains(BigInt(body.limitLoToHi))
        .grains(BigInt(body.limitHiToLo))
        .bytes(fromHex(body.signatureLo, 64))
        .bytes(fromHex(body.signatureHi, 64));
      break;
    }
    case "creditTransfer": {
      w.u8(7)
        .accountList(body.path.map(accountBytesFromJson))
        .grains(BigInt(body.amount))
        .bytes(fromHex(body.nonce, 16))
        .bytes(fromHex(body.signature, 64));
      break;
    }
    case "commonsSpend": {
      w.u8(8)
        .account(accountBytesFromJson(body.to))
        .grains(BigInt(body.amount))
        .str(body.policyRef);
      break;
    }
    default: {
      // Правило совместимости LIV.md §7: неизвестный вид записи — жёсткий отказ,
      // «мягкое непонимание» запрещено.
      const unknown: never = body;
      throw new EconomyCodecError(`неизвестный вид записи журнала: ${JSON.stringify(unknown)}`);
    }
  }
  return w.finish();
}

/** Хеш записи: `SHA-256(leaf-label ‖ seq ‖ at ‖ prev_hash ‖ body_bytes)`. */
export function entryHash(entry: LedgerEntryJson): Hash32 {
  const w = new CanonicalWriter();
  w.u64(entry.seq)
    .timestamp(entry.at)
    .hash32(fromHex(entry.prevHash, 32))
    .bytes(entryBodyCanonicalBytes(entry.body));
  return sha256Tagged(FEP_LEDGER_LEAF, w.finish());
}

/** Канонические байты head — вход подписи витнесса (`FEP_LEDGER_STH`). */
export function headCanonicalBytes(head: LedgerHeadJson): Uint8Array {
  const w = new CanonicalWriter();
  w.u64(head.size)
    .hash32(fromHex(head.lastEntryHash, 32))
    .hash32(fromHex(head.merkleRoot, 32))
    .timestamp(head.at);
  return w.finish();
}

// ---------- подписываемые байты транзакций (то, что подписывает кошелёк) ----------

/** Канонические байты авторизации перевода (подписывает отправитель). */
export function transferSigningBytes(
  from: Uint8Array,
  to: Uint8Array,
  amountGrains: bigint,
  nonce: Uint8Array,
): Uint8Array {
  if (nonce.length !== 16) {
    throw new EconomyCodecError(`nonce обязан быть 16 байт, получено ${nonce.length}`);
  }
  const w = new CanonicalWriter();
  w.account(from).account(to).grains(amountGrains).bytes(nonce);
  return w.finish();
}

/** Канонические байты авторизации линии доверия (подписывают обе стороны). */
export function trustlineSigningBytes(
  lo: Uint8Array,
  hi: Uint8Array,
  limitLoToHi: bigint,
  limitHiToLo: bigint,
): Uint8Array {
  const w = new CanonicalWriter();
  w.account(lo).account(hi).grains(limitLoToHi).grains(limitHiToLo);
  return w.finish();
}

/** Канонические байты авторизации платежа по цепочке доверия (подписывает плательщик). */
export function creditTransferSigningBytes(
  path: readonly Uint8Array[],
  amountGrains: bigint,
  nonce: Uint8Array,
): Uint8Array {
  if (nonce.length !== 16) {
    throw new EconomyCodecError(`nonce обязан быть 16 байт, получено ${nonce.length}`);
  }
  const w = new CanonicalWriter();
  w.accountList(path).grains(amountGrains).bytes(nonce);
  return w.finish();
}

// ---------- разбор JSON с границы системы ----------

function asRecord(value: unknown, what: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new EconomyCodecError(`${what}: ожидался JSON-объект`);
  }
  return value as Record<string, unknown>;
}

function readSafeInt(o: Record<string, unknown>, key: string, what: string): number {
  const v = o[key];
  if (typeof v !== "number" || !Number.isSafeInteger(v)) {
    throw new EconomyCodecError(`${what}.${key}: ожидалось целое число`);
  }
  return v;
}

function readStr(o: Record<string, unknown>, key: string, what: string): string {
  const v = o[key];
  if (typeof v !== "string") {
    throw new EconomyCodecError(`${what}.${key}: ожидалась строка`);
  }
  return v;
}

function readHex(o: Record<string, unknown>, key: string, bytes: number, what: string): string {
  const v = readStr(o, key, what);
  fromHex(v, bytes); // валидация формы; значение остаётся hex-строкой
  return v;
}

function readAccount(o: Record<string, unknown>, key: string, what: string): AccountIdJson {
  const v = o[key];
  if (!Array.isArray(v)) {
    throw new EconomyCodecError(`${what}.${key}: ожидался массив из 16 байт`);
  }
  const json = v as AccountIdJson;
  accountBytesFromJson(json); // валидация
  return json;
}

function parseParams(value: unknown, what: string): EconomyParametersJson {
  const o = asRecord(value, what);
  return {
    demurrage_ppm_per_period: readSafeInt(o, "demurrage_ppm_per_period", what),
    demurrage_period_ms: readSafeInt(o, "demurrage_period_ms", what),
    demurrage_exempt_threshold: readSafeInt(o, "demurrage_exempt_threshold", what),
    ubi_per_epoch: readSafeInt(o, "ubi_per_epoch", what),
    ubi_epoch_ms: readSafeInt(o, "ubi_epoch_ms", what),
    ubi_max_backfill_epochs: readSafeInt(o, "ubi_max_backfill_epochs", what),
    trustline_max_limit: readSafeInt(o, "trustline_max_limit", what),
    credit_path_max_hops: readSafeInt(o, "credit_path_max_hops", what),
  };
}

/** Строгий разбор head из JSON (граница системы: HTTP-ответ сервера). */
export function parseLedgerHead(value: unknown): LedgerHeadJson {
  const o = asRecord(value, "LedgerHead");
  return {
    size: readSafeInt(o, "size", "LedgerHead"),
    lastEntryHash: readHex(o, "lastEntryHash", 32, "LedgerHead"),
    merkleRoot: readHex(o, "merkleRoot", 32, "LedgerHead"),
    at: readSafeInt(o, "at", "LedgerHead"),
  };
}

function parseBody(value: unknown): LedgerEntryBodyJson {
  const o = asRecord(value, "EntryBody");
  const kind = readStr(o, "kind", "EntryBody");
  switch (kind) {
    case "genesis":
      return {
        kind,
        protocolVersion: readSafeInt(o, "protocolVersion", kind),
        params: parseParams(o.params, `${kind}.params`),
      };
    case "parametersUpdated":
      return {
        kind,
        params: parseParams(o.params, `${kind}.params`),
        policyRef: readStr(o, "policyRef", kind),
      };
    case "accountOpened":
      return {
        kind,
        account: readAccount(o, "account", kind),
        ownerKey: readHex(o, "ownerKey", 32, kind),
      };
    case "ubiIssued":
      return {
        kind,
        account: readAccount(o, "account", kind),
        fromEpoch: readSafeInt(o, "fromEpoch", kind),
        toEpoch: readSafeInt(o, "toEpoch", kind),
        amount: readSafeInt(o, "amount", kind),
      };
    case "demurrageCharged":
      return {
        kind,
        account: readAccount(o, "account", kind),
        periods: readSafeInt(o, "periods", kind),
        amount: readSafeInt(o, "amount", kind),
      };
    case "transfer":
      return {
        kind,
        from: readAccount(o, "from", kind),
        to: readAccount(o, "to", kind),
        amount: readSafeInt(o, "amount", kind),
        nonce: readHex(o, "nonce", 16, kind),
        signature: readHex(o, "signature", 64, kind),
      };
    case "trustlineSet":
      return {
        kind,
        lo: readAccount(o, "lo", kind),
        hi: readAccount(o, "hi", kind),
        limitLoToHi: readSafeInt(o, "limitLoToHi", kind),
        limitHiToLo: readSafeInt(o, "limitHiToLo", kind),
        signatureLo: readHex(o, "signatureLo", 64, kind),
        signatureHi: readHex(o, "signatureHi", 64, kind),
      };
    case "creditTransfer": {
      const pathRaw = o.path;
      if (!Array.isArray(pathRaw)) {
        throw new EconomyCodecError("creditTransfer.path: ожидался массив аккаунтов");
      }
      const path = pathRaw.map((item, i) => {
        if (!Array.isArray(item)) {
          throw new EconomyCodecError(`creditTransfer.path[${i}]: ожидался массив байт`);
        }
        accountBytesFromJson(item as AccountIdJson);
        return item as AccountIdJson;
      });
      return {
        kind,
        path,
        amount: readSafeInt(o, "amount", kind),
        nonce: readHex(o, "nonce", 16, kind),
        signature: readHex(o, "signature", 64, kind),
      };
    }
    case "commonsSpend":
      return {
        kind,
        to: readAccount(o, "to", kind),
        amount: readSafeInt(o, "amount", kind),
        policyRef: readStr(o, "policyRef", kind),
      };
    default:
      // Неизвестный вид записи — отказ целиком (LIV.md §7).
      throw new EconomyCodecError(`неизвестный вид записи журнала: ${kind}`);
  }
}

/** Строгий разбор записи журнала из JSON (граница системы). */
export function parseLedgerEntry(value: unknown): LedgerEntryJson {
  const o = asRecord(value, "LedgerEntry");
  return {
    seq: readSafeInt(o, "seq", "LedgerEntry"),
    at: readSafeInt(o, "at", "LedgerEntry"),
    prevHash: readHex(o, "prevHash", 32, "LedgerEntry"),
    body: parseBody(o.body),
  };
}
